### PR TITLE
Introduce `gosec` for Static Application Security Testing (SAST)

### LINKED ISSUE #948: Introduce gosec for Static Application Security Testing (SAST)
**How to categorize this issue?**

/area compliance
/area security
/kind enhancement
/priority 1

**What would you like to be added**:
use `gosec` for SAST. Check https://github.com/gardener/gardener/pull/9959 for reference.

- [x] mcm - https://github.com/gardener/machine-controller-manager/pull/954
- [x] mcm-provider-aws - https://github.com/gardener/machine-controller-manager-provider-aws/pull/179
- [x] mcm-provider-azure - https://github.com/gardener/machine-controller-manager-provider-azure/pull/169
- [x] mcm-provider-gcp - https://github.com/gardener/machine-controller-manager-provider-gcp/pull/134
- [x] mcm-provider-openstack - https://github.com/gardener/machine-controller-manager-provider-openstack/pull/199

**Why is this needed**:
Please look at the mail from Claudia with the title `Enable security linting in your repository by end of 2024` for details.

### PR DESCRIPTION
**Pull request by thiyyakat**
**What this PR does / why we need it**:

This PR introduces two make targets: `sast` and `sast-report` to run `gosec` for Static Application Security Testing. Additionally, it also addresses the security vulnerabilities in the MCM repository. It uses the default ruleset of gosec from gardener/gardener as introduced in https://github.com/gardener/gardener/pull/9959. 

`sast-report` target has also been added as a dependency for the `check` target.

**Which issue(s) this PR fixes**:
Partially fixes #948 

**Special notes for your reviewer**:
Tested manually to check if gosec is downloaded and installed to `hack/tools/bin`, if not already present, on running `make sast` and `make sast-report`.
Integration tests passed with provider-aws.

**Release note**:

```other developer
Adding `gosec` for Static Application Security Testing (SAST).
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @thiyyakat for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Review by rishabh-11 — APPROVED** [review]
/lgtm

File: pkg/apis/machine/validation/machinedeployment.go [hunk 1]
```diff
@@ -22,24 +23,49 @@ func internalValidateMachineDeployment(machineDeployment *machine.MachineDeploym
...
+	return true
+}
 
+func validateUpdateStrategy(spec *machine.MachineDeploymentSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
 	if spec.Strategy.Type != "RollingUpdate" && spec.Strategy.Type != "Recreate" {
 		allErrs = append(allErrs, field.Required(fldPath.Child("strategy.type"), "Type can either be RollingUpdate or Recreate"))
 	}
+	if spec.Strategy.Type == "RollingUpdate" {
+		if spec.Strategy.RollingUpdate == nil {
```
[the remark this answers is not in the document]
Comment (rishabh-11) [reply pkg/apis/machine/validation/machinedeployment.go:43]: The doc string is incorrect. If both are zero then we use `maxUnavailable` as 1.

File: pkg/apis/machine/validation/machinedeployment.go [hunk 2]
```diff
@@ -22,24 +23,49 @@ func internalValidateMachineDeployment(machineDeployment *machine.MachineDeploym
 	return allErrs
 }
 
-func validateMachineDeploymentSpec(spec *machine.MachineDeploymentSpec, fldPath *field.Path) field.ErrorList {
-	allErrs := field.ErrorList{}
-
-	if spec.Replicas < 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("replicas"), "Replicas has to be a whole number"))
+func canConvertIntOrStringToInt32(val *intstr.IntOrString, replicas int) bool {
```
[the remark this answers is not in the document]
Comment (rishabh-11) [reply pkg/apis/machine/validation/machinedeployment.go:26]: It's used here only. If we feel like it is required elsewhere, then we'll put it in separate location

**Pull request merged**